### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by JokeWaumans**
There are still some bugs that need to be fixed, but the main goal is working.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,22 @@
...
+
+Firstly, install the python packages with command:
+
+`pip install -r py-requirements.txt`
+
+Next, run the next command to set up a Neo4j database environment:
+
+`docker-compose up`
+
+Change the `BASE_URL` in the `config.yml` file to the correct path on you're computer.
```
Comment (Ben2022) [inline README.md]:
```suggestion
Change the `BASE_URL` in the `config.yml` file to the correct path on your computer.
```

File: autocomplete.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+"""
+Python module that is needed for the autocompletion.
+Copyright (c) 2022 Melexis
+"""
+
+from neo4j import GraphDatabase
+import json
+from ruamel.yaml import YAML
+
```
Comment (Ben2022) [inline autocomplete.py:10]: Generally speaking, I think we need more comments explaining what we do.

File: make_html.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
+"""
+Python module to make the html files of the traceabilitytool with the accompanying data to start with.
```
Comment (Ben2022) [inline make_html.py]:
```suggestion
Python module to make the html files of the traceability tool with the accompanying data to start with.
```

File: static/autocomplete.js [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+                inp.value = this.getElementsByTagName("input")[0].value;
+                /* close the list of autocompleted values,
+                (or any other open lists of autocompleted values */
+                closeAllLists();
+            });
+            a.appendChild(b);
+          }
+      }
+  });
+  /* execute a function presses a key on the keyboard */
```
Comment (Ben2022) [inline static/autocomplete.js]: Can you rephrase this, please?

File: static/initialize.js [hunk 5]
```diff
@@ -0,0 +1,110 @@
+/**
+ * @fileoverview This file contains variables that are needed for the traceablility vieuwer.
```
Comment (Ben2022) [inline static/initialize.js]:
```suggestion
 * @fileoverview This file contains variables that are needed for the traceability viewer.
```

File: static/initialize.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+for (let [key,value] of Object.entries(config["group_colours"])){
+    legend_groups.push(key);
+    legend_colours.push(value);
+}
+
+legend_groups.push("Others");
+legend_colours.push("#000000");
+
+// for (let [key,value] of Object.entries(config["link_colours"])){
```
Comment (Ben2022) [inline static/initialize.js]: Can we clean that up?

File: static/initialize.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
+/**
```
Comment (Ben2022) [inline static/initialize.js:1]: Generally speaking, I think we need more comments explaining what we do.

File: static/legend.js [hunk 8]
```diff
@@ -0,0 +1,253 @@
+/**
```
Comment (Ben2022) [inline static/legend.js:1]: The structure of the file is not clear to me. Is it only me or you share the same concern, @JasperCraeghs ?

File: static/search.js [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+    alert("The search input field is empty.")
+    clearInterval(load);
+  }
+  else {
+    search_update(search);
+  }
+});
+
+/**
+ * This function define a query for showing the V-model of the given ID.
```
Comment (Ben2022) [inline static/search.js]:
```suggestion
 * This function defines a query for showing the V-model of the given ID.
```

**Comment by Ben2022** [comment]
Is it possible that we perform some unit tests in general? I will be more explicit in my next comment.

File: static/search.js [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+    clearInterval(load);
+    alert("For this item, this feature is not implemented.\nTry another item.")
+  }
+}
+
+/**
+ * This function will give the result of the query that is defined for the search input.
+ * @param {string} query The defined query
+ */
+function addConnectedNodes(query){
```
Comment (Ben2022) [inline static/search.js]: Do we have evidence that it works as we expect?

Comment (JokeWaumans) [reply static/search.js]: Evidence of what type? If the graph database is correctly built?

Comment (Ben2022) [reply static/search.js]: Have we tested it and if yes, can we commit also the tests?

Comment (JokeWaumans) [reply static/search.js]: It works on the local build of a product, so there are no real tests.

Comment (Ben2022) [reply static/search.js]: @JasperCraeghs , do we have unit tests for the traceability plugin?

File: autocomplete.py (hunk 2 above)
Comment (JokeWaumans) [reply autocomplete.py:10]: I think it is resolved in 9834004

File: static/search.js [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+        .catch(function(err){
+          alert("Error:" + err);
+          clearInterval(load);
+        });
+}
+
+
+
+/**
+ * This function will update the graph with the new nodes and links. It will restart the simulation.
```
Comment (JasperCraeghs) [inline static/search.js]:
```suggestion
 * Updates the nodes and links of the graph and restarts the simulation.
```
Please update every docstring to use this style of wording.

File: static/vis_D3.css [hunk 12]
```diff
@@ -0,0 +1,128 @@
+* {
+  box-sizing: border-box;
+}
+
+body {
+  font: 16px Arial;
```
Comment (JasperCraeghs) [inline static/vis_D3.css]:
```suggestion
  font-size: 16px;
  font-family: Lato, Helvetica Neue, Arial, sans-serif;
```
Let's use the same font-family as sphinx-rtd-theme.

File: autocomplete.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
+"""
+Python module that is needed for the autocompletion. One is for the query input field that contains all node IDs,
+links and words that are often used in a Cypher query. The second one is for the search input field.
+This contains all the node IDs where the group is a main group of the V model (not the group "Others").
+Copyright (c) 2022 Melexis
```
Comment (JasperCraeghs) [inline autocomplete.py]: This copyright notice should be removed as we rely on the [`LICENSE`](https://github.com/melexis/traceability-viewer/blob/main/LICENSE) file, right @Letme ?

File: config.yml [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+group_colours:
+  SYSRQT: "#fc851c"
+  SWRQT: "#FF33CC"
+  SWARCH: "#7777FF"
+  DESIGN: "#cf5402"
+  QTEST: "#00FF00"
+  SWITEST: "#00FF00"
+  UTEST: "#00FF00"
+
+link_colours:
```
Comment (JasperCraeghs) [inline config.yml]:
```suggestion
link_colors:
```
Let's use American English. Please update all other occurrences. Or does a  dependency like D3js use "colour" everywhere?

File: autocomplete.py (hunk 13 above)
Comment (Letme) [reply autocomplete.py]: you want a single LICENSE file per repository and nothing on top of the files. Also make sure its 2023, not 2022.

File: autocomplete.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+"""
+
+import json
+from neo4j import GraphDatabase
+from ruamel.yaml import YAML
+
+# open the config file
+with open("config.yml", "r", encoding="utf-8") as open_file:
```
Comment (Letme) [inline autocomplete.py:12]: Not sure why you do not make a function here and just have all the code/execution in the single file. Use functions (`def main:` if nothing else)

File: make_html.py [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+            prop["label"] = record["r"].type
+            links.append(prop)
+
+        if record["m"]:
+            if record["m"]._properties not in nodes:
+                nodes.append(record["m"]._properties)
+
+    graph = {"nodes": nodes, "links": links}
+
+    html = f"""
```
Comment (Letme) [inline make_html.py:29]: I suggest you use mako template for such HTML blobs. It will make it easier for future generations to change and it will not add so many tabs infront.

File: config.yml [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+  configurable_by: "#001ea3"
+  implements: "#b1ba02"
+  implemented_by: "#6f0182"
+  realizes: "#499161"
+  realized_by: "#820101"
+  fulfills: "#ff9b05"
+  validates: "#277a7a"
+  validated_by: "#27597a"
+  passed_by: "#8c00ff"
+  passes: "#2b00ff"
```
Comment (JasperCraeghs) [inline config.yml:44]: Did you use a random color picker tool? I suggest using green for passes/passed_by and red for fails/failed_by and orange for skiped/skipped_by. These relationships are associated with specific colors. For other relationships, the colors matters less I would say.

File: make_html.py [hunk 18]
```diff
@@ -0,0 +1,168 @@
+"""
+Python module to make the html files of the traceability tool with the accompanying data to start with.
+Copyright (c) 2022 Melexis
+"""
+
+from autocomplete import graphdb, words, search_ids, config, config_dict
+
+with graphdb.session() as session:
```
Comment (Letme) [inline make_html.py:7]: Again, make sure you use functions.

File: config.yml [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+  guaranteed_by: "#ffdd00"
+  guaranteed: "#ff6f00"
+  ext_polarion_reference: "#01173b"
+  test_setup_used: "#006699"
+  is_attribute_of: "#009900"
```
Comment (JasperCraeghs) [inline config.yml:49]: Did you implement a warning when the traceable collection contains a relationship that is not configured here?

File: static/autocomplete.js (hunk 4 above)
Comment (JokeWaumans) [reply static/autocomplete.js]: Resolved in 21e0e1ffc0022ce7df3de6db9a3677760fc1e691

File: static/initialize.js (hunk 7 above)
Comment (JokeWaumans) [reply static/initialize.js:1]: Resolved in 7d3ac97116fba4f7045d79a4d5615638dc7a38e0

File: autocomplete.py (hunk 13 above)
Comment (JokeWaumans) [reply autocomplete.py]: Resolved in 82c6d6cb426b2a0c6b57a23fa560c40329732620

File: config.yml (hunk 14 above)
Comment (JokeWaumans) [reply config.yml]: Resolved in 83a206dcb8dc459fb807e351a7600c2d92024b89

File: make_neo4j_db.py [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+                    session.execute_write(self.add_item, id1)
+                    nodes_made.append(id1)
+
+                if targets != {}:
+                    for rel in targets:
+                        for id2 in targets[rel]:
+                            if id2 not in nodes_made:
+                                session.execute_write(self.add_item, id2)
+                                nodes_made.append(id2)
+                            if rel in list(self.config_dict["link_colors"].keys()):
```
Comment (JasperCraeghs) [inline make_neo4j_db.py:73]:
```
color = self.config_dict["link_colors"].get(rel, "#808080")
session.execute_write(self.add_relationship, id1, id2, rel, color)
```

File: make_neo4j_db.py [hunk 21]
```diff
@@ -0,0 +1,212 @@
...
+        """Make the Neo4j database."""
+        with self.driver.session() as session:
+            session.run(
+                """MATCH (n)
+                            DETACH DELETE n"""
+            )
+            nodes_made = []
+            for i, item in enumerate(self.data):
+                targets = item["targets"]
+                attr = item["attributes"]
```
Comment (JasperCraeghs) [inline make_neo4j_db.py:59]: `attributes`

File: make_neo4j_db.py [hunk 22]
```diff
@@ -0,0 +1,212 @@
...
+            id=id_item,
+            props=properties,
+        )
+
+    @staticmethod
+    def add_attributes(tx, id_item, attributes):
+        """Add the attributes of the document item in the graph database.
+        Args:
+            id_item (str): ID of the document item
+            attributes (json): JSON string of the attributes of the document item
```
Comment (JasperCraeghs) [inline make_neo4j_db.py]:
```suggestion
            attributes (str): JSON string of the attributes of the document item
```

File: make_neo4j_db.py [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+        tx.run(
+            """MATCH (n:DocItem)
+            WHERE n.id = $id
+            SET n.attributes = $attr""",
+            id=id_item,
+            attr=attributes,
+        )
+
+    @staticmethod
+    def add_group_and_color(tx, id_item, group, color):
```
Comment (JasperCraeghs) [inline make_neo4j_db.py:154]: add_item_to_group

File: make_neo4j_db.py [hunk 24]
```diff
@@ -0,0 +1,212 @@
...
+                                    self.add_relationship, id1, id2, rel, self.config_dict["link_colors"][rel]
+                                )
+                            else:
+                                session.execute_write(self.add_relationship, id1, id2, rel, "#808080")
+                del props["targets"]
+                session.execute_write(self.add_attributes, id1, json.dumps(attr))
+                del props["attributes"]
+                session.execute_write(self.add_properties, id1, props)
+
+                if id1.startswith(self.groups[0]):
```
Comment (JasperCraeghs) [inline make_neo4j_db.py:84]:
```
for group in groups:
    if re.match(group, id1):
        selected_group = group
        selected_color = self.config_dict["group_colors"][group]
        break
else:
    selected_group = "Others"
    selected_color = "#000000"
session.execute_write(self.add_group_and_color, id1, selected_group, selected_color)
```
Let's use [`re.match`](https://docs.python.org/3/library/re.html#re.match) instead of `str.startswith` for added flexibility. Users of mlx.traceability are used to defining regular expressions. Please document this feature.

File: README.md [hunk 25]
```diff
@@ -1 +1,57 @@
...
+
+Firstly, install the python packages with command:
+
+`pip install -r py-requirements.txt`
+
+Next, run the next command to set up a Neo4j database environment:
+
+`docker-compose up`
+
+If docker-compose is not istalled yet, install it with the `sudo apt install docker-compose` command.
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
If docker-compose is not installed yet, install it with the `sudo apt install docker-compose` command.
```
Please provide a link to the official documentation of Docker Compose and any other tool, e.g. Neo4j.

File: README.md [hunk 26]
```diff
@@ -1 +1,57 @@
-# traceability-viewer
\ No newline at end of file
+# traceability-viewer
+
+## Get started
+
+Firstly, install the python packages with command:
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
Firstly, install the Python dependencies with [pip](https://pip.pypa.io/en/stable/):
```

**Comment by JasperCraeghs** [comment]
> There are still some bugs that need to be fixed, but the main goal is working.

Please document the list of known bugs.

File: README.md [hunk 27]
```diff
@@ -1 +1,57 @@
...
+The path to the html documentation. Depending on the file structure, the ending could be changed to `/rom/html/document.html#id` in case of ROM.
+
+# Filters
+
+The groups of the V-model. This is in order, beginning at the left side of the V-mdoel: from top to bottom and the right side (the test side) also from top to bottom.
+These groups are also the filters.
+
+# Group colors
+
+The groups with their corresponding colors.
```
Comment (JasperCraeghs) [inline README.md:59]: Can we combine the `filters` and `group_colors` in the configuration into a single map/dictionary? Python>=3.6 [preserves the order of dictionaries](https://stackoverflow.com/questions/1867861/how-to-keep-keys-values-in-same-order-as-declared). Does this work when reading a YAML map, though?

File: config.yml [hunk 28]
```diff
@@ -0,0 +1,50 @@
+---
+variables:
+  PRODUCT: &PRODUCT 90427
+  BASE_URL: &BASE_URL /path/to/product/build/doc/    # CHANGE THIS
+
+project: MLX{product}
```
Comment (JasperCraeghs) [inline config.yml:6]: Is this configuration value used by the tool? If so, document it in the readme.

File: README.md [hunk 29]
```diff
@@ -1 +1,63 @@
...
+
+`docker-compose up`
+
+If docker-compose is not istalled yet, install it with the `sudo apt install docker-compose` command.
+To fix issues with permission, run the code below:
+`sudo chmod 666 /var/run/docker.sock`
+`sudo setfacl --modify user::rw /var/run/docker.sock`
+
+Change the `BASE_URL` in the `config.yml` file to the correct path on your computer.
+This is the path to the JSON database of a local build (doc folder) of a Melexis product.
```
Comment (JasperCraeghs) [inline README.md:28]: This will be an open source tool, so it will be available to every company. The example config.yml can contain some Melexis-specific configuration, but the documentation for chapter Configuration should be exhaustive and generic enough for everyone. I suggest simply referencing the Configuration section here.

`Edit the example [configuration file](#configuration).`

File: config.yml [hunk 30]
```diff
@@ -0,0 +1,50 @@
...
+
+json_folder: [*BASE_URL, *PRODUCT, /database.json]
```
Comment (JasperCraeghs) [inline config.yml:8]: I would rename this configuration value to `json_database` or `traceability_export`. 

`/database.json` should not contain a `/`. Then, every element of the list is a part of the path, which you can assemble with [`pathlib.PurePosixPath.joinpath`](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.joinpath).

File: README.md [hunk 31]
```diff
@@ -1 +1,63 @@
...
+
+Finally, you can open one of the HTML files and start having fun.
+
+## Configuration
+
+The configuration takes the form of a YAML file.
+
+### Variables
+
+The product number and the BASE_URL contains the path to the build/doc folder of that product.
```
Comment (JasperCraeghs) [inline README.md:41]: Please explain the syntax for defining and using variables.

Support for environment variables is needed for easy integration in CI.

File: config.yml [hunk 32]
```diff
@@ -0,0 +1,50 @@
...
+# URL: [*BASE_URL, *PRODUCT, /rom/html/document.html#id]
+URL: [*BASE_URL, *PRODUCT, /html/document.html#id]
```
Comment (JasperCraeghs) [inline config.yml]:
```suggestion
html_dir: [*BASE_URL, *PRODUCT, html]
```
`document.html#id` is unnecessary in the configuration. `node_clicked` in forcegraph.js can append the document name and node ID to `html_dir`.

File: README.md [hunk 33]
```diff
@@ -1 +1,63 @@
...
+
+### JSON folder
+
+The path to the exported database.json file. This is the BASE_URL followed by the product number, ending with /database.json.
+
+### URL
+
+The path to the HTML documentation. Depending on the file structure, the ending could be changed to `/rom/html/document.html#id` in case of ROM.
```
Comment (JasperCraeghs) [inline README.md:49]: Please clarify that you mean the HTML documentation generated by mlx.traceability, not `make_html.py`.

Is a URL to documentation hosted on a webserver supported?

The `rom` subfolder is Melexis-specific and I don't think we will use this. I would not mention it here.

File: autocomplete.py [hunk 34]
```diff
@@ -0,0 +1,61 @@
...
+                    if record["n"]._properties["id"].startswith(group):
+                        search_ids.append(record["n"]._properties["id"])
+        # save the types of links
+        if record["r"]:
+            if record["r"].type not in link_types:
+                link_types.append(record["r"].type)
+
+# add the words of a query that are used the most.
+for w in ["MATCH", "STARTS WITH", "CONTAINS", "WHERE", "RETURN"]:
+    words.append(w)
```
Comment (JasperCraeghs) [inline autocomplete.py:58]: You could use this list as the initial value instead of `words = []` on L28. Or does the order of the list elements matter?

File: autocomplete.py [hunk 35]
```diff
@@ -0,0 +1,61 @@
...
+
+for l in link_types:
+    words.append(l)
```
Comment (JasperCraeghs) [inline autocomplete.py:61]: Use [`list.extend`](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists)

File: README.md [hunk 36]
```diff
@@ -1 +1,63 @@
...
+
+The configuration takes the form of a YAML file.
+
+### Variables
+
+The product number and the BASE_URL contains the path to the build/doc folder of that product.
+
+### JSON folder
+
+The path to the exported database.json file. This is the BASE_URL followed by the product number, ending with /database.json.
```
Comment (JasperCraeghs) [inline README.md:45]: In addition to support for a local file, we should support the download of this file when it's hosted on a webserver. See https://docs.python.org/3/library/urllib.parse.html#url-parsing

**Review by JasperCraeghs — CHANGES_REQUESTED** [review]

File: make_html.py [hunk 37]
```diff
@@ -0,0 +1,167 @@
...
+            <label>Cypher query: </label>
+            <div class="autocomplete" style="width:100%;">
+                <input placeholder="MATCH (source)<-[rel]->(target) WHERE source.id = 'string_id' RETURN source, rel, target" type="text" id="query" value="MATCH (source)<-[rel]->(target) WHERE source.id CONTAINS 'MEMORY' RETURN source,rel,target">
+                <button class="btn" id="query_button">Submit query</button>
+            </div>
+            <!-- <input placeholder="MATCH (source)-[rel]->(target) WHERE source.id = 'string_id' RETURN source, rel, target" margin-left="100px" size="80%" type="text" id="query" value="MATCH (source)-[rel]->(target) WHERE source.id CONTAINS 'MEMORY' RETURN source,rel,target"> -->
+            <br>
+            <label>Search: </label>
+            <div class="autocomplete" style="width:100%;">
+                <input placeholder="SWARCH_" type="text" id="search">
```
Comment (JasperCraeghs) [inline make_html.py:56]: `SWARCH_` is Melexis-specific. What's the value of a placeholder here?

File: static/initialize.js [hunk 38]
```diff
@@ -0,0 +1,126 @@
...
+let page = path.split("/").pop().split(".")[0];
+console.log(page);
+
+// variables for the width and height
+let width = window.innerWidth - 20;
+let height = window.innerHeight - 120;
+
+// the links that will be visible
+let filtered_links = ["impacts_on", "fulfilled_by", "validates", "configurable_by", "realized_by", "passed_by",
+  "guaranteed_by", "ext_polarion_reference", "test_setup_used", "is_attribute_of"];
```
Comment (JasperCraeghs) [inline static/initialize.js:18]: This must be configurable.

File: config.yml (hunk 19 above)
Comment (JokeWaumans) [reply config.yml:49]: No, when a relationship is not configured, it will get a gray color.

File: config.yml (hunk 28 above)
Comment (JokeWaumans) [reply config.yml:6]: `project: MLX{product}` is not used, thus can be removed

File: make_html.py (hunk 37 above)
Comment (JokeWaumans) [reply make_html.py:56]: Yes, it is Melexis-specific, and it is no longer relevant because you need to enter a full ID.

File: static/search.js (hunk 10 above)
Comment (JasperCraeghs) [reply static/search.js]: The traceability plugin has some unit tests. We rely a lot on the build of the example documentation. If a new feature is added, it is showcased in the example doc and the HTML is inspected manually.

File: config.yml (hunk 17 above)
Comment (JokeWaumans) [reply config.yml:44]: Changed the colors in 495818dd280896df70bbd32ab1bd852ad60fcd85

File: make_neo4j_db.py (hunk 21 above)
Comment (JokeWaumans) [reply make_neo4j_db.py:59]: Resolved in c44e254bb241c8a208f74b8312301350afdbfec8

File: README.md (hunk 27 above)
Comment (JokeWaumans) [reply README.md:59]: Not relevant anymore, see #2

File: make_html.py (hunk 16 above)
Comment (JasperCraeghs) [reply make_html.py:29]: Resolved in #2 by using Vue components

**Comment by JasperCraeghs** [comment]
Closing in favor of #2, which uses Django for the backend and Vue for the frontend.
